### PR TITLE
SONAR-6817 Issues mode should support analysis of projects not associ…

### PR DESCRIPTION
**Pull request by dbmeneses**
Most important classes changed: NonAssociatedCacheSynchronizer and ProjectReactorBuilder

### CODE REVIEW

File: sonar-batch/src/main/java/org/sonar/batch/scan/ProjectScanContainer.java [hunk 1]
```diff
@@ -190,9 +201,16 @@ private void addBatchComponents() {
...
-    addIfMissing(DefaultProjectRepositoriesLoader.class, ProjectRepositoriesLoader.class);
     addIfMissing(DefaultServerIssuesLoader.class, ServerIssuesLoader.class);
     addIfMissing(DefaultServerLineHashesLoader.class, ServerLineHashesLoader.class);
+    addIfMissing(DefaultActiveRulesLoader.class, ActiveRulesLoader.class);
+    addIfMissing(DefaultQualityProfileLoader.class, QualityProfileLoader.class);
+    addIfMissing(DefaultProjectRepositoriesLoader.class, ProjectRepositoriesLoader.class);
+    addIfMissing(DefaultProjectSettingsLoader.class, ProjectSettingsLoader.class);
+  }
+
+  private boolean isProjectAssociated() {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/scan/ProjectScanContainer.java]: MAJOR Private method 'isProjectAssociated' is never used. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 6 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/input' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/input) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/output' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/output) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MINOR [Remove this unused import 'org.sonar.batch.rule.ModuleQProfiles'.](https://github.com/SonarSource/sonarqube/blob/8b95fb5c444b22498695493e88afb91081806e11/sonar-batch/src/main/java/org/sonar/batch/scan/ModuleScanContainer.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/db/protobuf' package (org.codehaus.sonar:sonar-db:src/main/gen-java/org/sonar/db/protobuf) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonarqube/ws' package (org.codehaus.sonar:sonar-ws:src/main/gen-java/org/sonarqube/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

File: sonar-batch/src/main/java/org/sonar/batch/analysis/DefaultAnalysisMode.java [hunk 2]
```diff
@@ -77,6 +83,9 @@ public void printMode() {
     if (mediumTestMode) {
       LOG.info("Medium test mode");
     }
+    if (notAssociated) {
+      LOG.info("Project is not associated with the server");
```
Comment (henryju) [inline sonar-batch/src/main/java/org/sonar/batch/analysis/DefaultAnalysisMode.java]: As we focus this mode for users having no knowledge of SQ server I would rephrase this log to something like: "Local analysis". WDYT @bellingard ?

Comment (bellingard) [inline sonar-batch/src/main/java/org/sonar/batch/analysis/DefaultAnalysisMode.java]: @henryju Agree.

File: sonar-batch/src/main/java/org/sonar/batch/bootstrap/GlobalContainer.java [hunk 3]
```diff
@@ -123,14 +123,22 @@ private void installPlugins() {
...
+      String projectKey = getProjectKeyWithBranch(props);
+      new ProjectSyncContainer(this, projectKey, false).execute();
     }
     new ProjectScanContainer(this, props, components).execute();
   }
 
-  public void syncProject(Map<String, String> analysisProperties, boolean force) {
-    AnalysisProperties props = new AnalysisProperties(analysisProperties, this.getComponentByType(GlobalProperties.class).property(CoreProperties.ENCRYPTION_SECRET_KEY_PATH));
-    new ProjectSyncContainer(this, props, force).execute();
+  private static String getProjectKeyWithBranch(AnalysisProperties props) {
```
Comment (henryju) [inline sonar-batch/src/main/java/org/sonar/batch/bootstrap/GlobalContainer.java]: @CheckForNull ?

File: sonar-batch/src/main/java/org/sonar/batch/cache/ProjectSyncContainer.java [hunk 4]
```diff
@@ -19,59 +19,73 @@
...
   public void doAfterStart() {
-    getComponentByType(ProjectCacheSynchronizer.class).load(force);
+    if (projectKey != null) {
+      getComponentByType(ProjectCacheSynchronizer.class).load(projectKey, force);
+    } else {
+      getComponentByType(NonAssociatedCacheSynchronizer.class).execute(force);
+    }
+  }
+
+  private static DefaultAnalysisMode createIssuesAnalisysMode() {
```
Comment (henryju) [inline sonar-batch/src/main/java/org/sonar/batch/cache/ProjectSyncContainer.java]: s/createIssuesAnalisysMode/createIssuesAnalysisMode

**Comment by henryju** [comment]
Appart previous comments, LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 5 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/input' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/input) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/output' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/output) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MINOR [Remove this unused import 'org.sonar.batch.rule.ModuleQProfiles'.](https://github.com/SonarSource/sonarqube/blob/78c1508afb8eb6e6deffbc9f093acf34c869d20f/sonar-batch/src/main/java/org/sonar/batch/scan/ModuleScanContainer.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/db/protobuf' package (org.codehaus.sonar:sonar-db:src/main/gen-java/org/sonar/db/protobuf) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonarqube/ws' package (org.codehaus.sonar:sonar-ws:src/main/gen-java/org/sonarqube/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 5 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/input' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/input) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/output' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/output) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MINOR [Remove this unused import 'org.sonar.batch.rule.ModuleQProfiles'.](https://github.com/SonarSource/sonarqube/blob/1d99d78844ccc3efe9400b89d3bd73f1c1c0404a/sonar-batch/src/main/java/org/sonar/batch/scan/ModuleScanContainer.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/db/protobuf' package (org.codehaus.sonar:sonar-db:src/main/gen-java/org/sonar/db/protobuf) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonarqube/ws' package (org.codehaus.sonar:sonar-ws:src/main/gen-java/org/sonarqube/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 5 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/input' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/input) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/output' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/output) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MINOR [Remove this unused import 'org.sonar.batch.rule.ModuleQProfiles'.](https://github.com/SonarSource/sonarqube/blob/1d99d78844ccc3efe9400b89d3bd73f1c1c0404a/sonar-batch/src/main/java/org/sonar/batch/scan/ModuleScanContainer.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/db/protobuf' package (org.codehaus.sonar:sonar-db:src/main/gen-java/org/sonar/db/protobuf) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonarqube/ws' package (org.codehaus.sonar:sonar-ws:src/main/gen-java/org/sonarqube/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
